freckle-app/library/Freckle/App/TextualEnum.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | Typical instances for enumerated data types with textual representation
+module Freckle.App.TextualEnum
+  ( TextualEnum (..)
+  , EnumValue (..)
+  , EnumName (..)
+  , enums
+  , fromText
+  , prop_roundTripEnumText
+  ) where
+
+import Freckle.App.Prelude
+
+import Autodocodec
+import Control.Lens hiding (elements)
+import Control.Monad (mzero)
+import Data.Aeson
+import Data.Csv qualified as CSV
+import Data.Dynamic (Typeable)
+import Data.List.NonEmpty qualified as NE
+import Data.OpenApi
+import Data.Text.Encoding qualified as T
+import Database.Persist.Sql
+  ( PersistField (..)
+  , PersistFieldSql (..)
+  , SqlType (..)
+  )
+import Servant
+import Test.QuickCheck (Arbitrary (..), elements)
+import Web.PathPieces
+
+class EnumValue a where
+  -- | Convert a 'TextualEnum' to 'Text'
+  toText :: a -> Text
```

Comment:
A thought -

```haskell
data TextMapping a =
  TextMapping !(a -> Text) !(Text -> Maybe a)

class EnumValue a where
  textMapping :: TextMapping a

enumBoundedTextMapping :: (Enum a, Bounded a) => (a -> Text) -> TextMapping a
enumBoundedTextMapping f =
  let
    !m = Map.fromList $ (id &&& f) <$> [minBound..maxBound]
  in
    TextMapping f (flip (Map.lookup m))
```

```haskell
data ContentArea = Math | Ela
  deriving (...) via TextualMapping ContentArea

instance EnumValue ContentArea where
  textMapping = enumBoundedTextMapping $ \case
    Math -> "math"
    Ela -> "ela"
```

This approach does two things -

- Allows the user options other than enum/bounded to enumerate if they want
- Allows a logarithmic rather than linear lookup from a `Map` which is constructed only once (not 100% sure if that would require explicitly lifting each `textMapping` definition to the top level or not, would want a benchmark to test.)

Reply:
I thought about doing something like this but I came to the conclusion that basically
> not 100% sure if that would require explicitly lifting each textMapping definition to the top level or not

was kind of hazy (i.e. I'm unsure exactly how to get GHC to compute it once) and thus prone to being incorrectly done.

I'm also not convinced that a hash map is actually quicker for smaller enums. Hashing a value vs. looking through like 4-13 cases feels like a wash. I could be totally wrong here as I have no data to back this up but the tradeoffs don't feel worth it at first blush.

Reply:
personally I greatly prefer using deriving via instead of wrapping a bunch of types with a utility type; I'm fine with wrapping in a utility type for calculation purposes, but it gets really stinky when someone does e.g.

```
data Person = { age :: Sum Int } 

totalAge :: [Person] -> Int
totalAge = getSum .  fold . age 
```
instead of, like:
```
data Person = { age :: Int } 

totalAge :: [Person] -> Int
totalAge = getSum . fold <$> Sum . age 

-- which would let you then do

youngestAge :: [Person] -> Int
youngestAge = getMin . fold <$> Min . age 
```
.. tho I admit ive seen it in the our codebase a few times, and that this is a personal preference that I can't justify beyond "confuses the data model" and "at what point do you stop adding utility wrappers"

Tho, lol, I guess this utility wrapper thing is kinda sorta analogous to monad transformer stacks, eh?

I don't really have an opinion on the linear vs log search. If the N is small it doesn't matter much. I guess if someone did the following, it would matter a lot:

```
newtype Age = Age Int
  deriving (...) via TextualMapping Age

instance EnumValue ContentArea where
  textMapping = enumBoundedTextMapping show

main = do putStrLn $ show $ Age maxBound
```
but idk this seems like quite the goofy thing to do, and hey, I notice that you address it. If you do it via mapping, and someone were to enumerate the set of Ints, you'd end up eating up so much memory... so idk. 

you could make a way for the user to use a second function tho so a body could pick, e.g. `enumBoundedTextMapping` and `enumBoundedTextMappingMemo` or something

Reply:
> greatly prefer using deriving via instead of wrapping a bunch of types with a utility type

Agreed and I'm wondering if maybe the spec here would be a good place to demonstrate using `TextualEnum` to define instances directly on the `PrimaryColor` type.